### PR TITLE
Use generators for Worker Activity report in email [V2]

### PR DESCRIPTION
**Pull request by pr33thi**
Here's the new and improved version of [this PR](https://github.com/dimagi/commcare-hq/pull/23734)
I took your comments into account and made a few changes, generators should be used up until the last step where it gets written to excel.
Note, I am still testing this in a limited release to make sure it works at ICDS scale, if it doesn't I may need to add updates to this, but it takes so long to run that I wanted to get a review sooner rather than later.

[Asana task](https://app.asana.com/0/705279448066644/1108792991464102/f)
Description copy pasted from the above PR: 
The issue is that the report used to pull the entire table from ES and keep it in memory, format it, and then write it to excel, which would raise an ES error because the table is too large for ICDS. This uses all generators instead of ever saving the whole table in memory.

**Also**: Should this be QA'd? I tested what I could locally but am not confident I got every edge case.

@calellowitz 
@esoergel @sravfeyn

### CODE REVIEW

File: corehq/ex-submodules/couchexport/export.py [hunk 1]
```diff
@@ -313,19 +316,28 @@ def get_data(self):
...
-                (name, [cls(row) for row in rows])
-            )
-        return ret
+            return rows
+
+        wrapped_rows = []
+        for name, rows in tables:
+            if isinstance(rows, itertools.chain):
+                first_entry_rows, rows_to_wrap = itertools.tee(rows)
+                first_entry = first_entry_rows.next()
```
[the remark this answers is not in the document]
Comment (nickpell) [reply corehq/ex-submodules/couchexport/export.py]: stickler wants you to use the builtin ```next``` function:
https://docs.python.org/3/library/functions.html#next

File: corehq/apps/reports/standard/monitoring.py [hunk 2]
```diff
@@ -1792,22 +1801,74 @@ def _total_row(self, rows, report_data):
...
+        )
+        chunk_size = 10000
+        first_id_in_chunk = 0
+        user_chunk_total = user_es_query.fields(util.SimplifiedUserInfo.ES_FIELDS).size(chunk_size).run().total
+        while first_id_in_chunk < user_chunk_total:
+            user_chunk = user_es_query.fields(util.SimplifiedUserInfo.ES_FIELDS).start(first_id_in_chunk).size(chunk_size).run().hits
+            users = [util._report_user_dict(user) for user in user_chunk]
+            formatted_data = self._report_data(users_to_iterate=users)
+            if self.view_by_groups:
+                rows = self._rows_by_group(formatted_data)
```
Comment (esoergel) [inline corehq/apps/reports/standard/monitoring.py]: I don't think this'll work, since its unlikely all users in a group are going to be in a single chunk.  I'd try to make this feature unavailable for `view_by_groups`.

File: corehq/apps/reports/standard/monitoring.py [hunk 3]
```diff
@@ -1792,22 +1801,74 @@ def _total_row(self, rows, report_data):
...
+            else:
+                rows = self._rows_by_user(formatted_data, users)
+            this_row = self._total_row(rows, formatted_data)
```
Comment (esoergel) [inline corehq/apps/reports/standard/monitoring.py]: nit: I'd call this `partial_total` or something

File: corehq/apps/reports/standard/monitoring.py [hunk 4]
```diff
@@ -1792,22 +1801,74 @@ def _total_row(self, rows, report_data):
...
+            self.total_row = self.sum_rows_together(self.total_row, this_row)
+            for row in rows:
+                yield row
+            first_id_in_chunk = first_id_in_chunk + chunk_size
+        self.total_row = self.format_total_row(self.total_row)
+        yield self.total_row
```
Comment (esoergel) [inline corehq/apps/reports/standard/monitoring.py]: I like this approach for the total row

File: corehq/apps/reports/standard/monitoring.py [hunk 5]
```diff
@@ -1792,22 +1801,74 @@ def _total_row(self, rows, report_data):
...
+
+    @staticmethod
+    def format_total_row(unformatted_total_row):
```
Comment (esoergel) [inline corehq/apps/reports/standard/monitoring.py]: nit: I'd mark this method and `sum_rows_together` private

File: corehq/apps/saved_reports/tasks.py [hunk 6]
```diff
@@ -145,20 +147,29 @@ def send_email_report(self, recipient_emails, domain, report_slug, report_type,
...
+            message="Encountered error while sending email",
+            details={
+                'subject': subject,
+                'recipients': str(recipient_emails),
+                'error': er,
+            }
+        )
+        if getattr(er, 'smtp_code', None) in LARGE_FILE_SIZE_ERROR_CODES or type(er) == ESError:
             # If the smtp server rejects the email because of its large size.
             # Then sends the report download link in the email.
```
Comment (esoergel) [inline corehq/apps/saved_reports/tasks.py]: Can you update this comment?

File: corehq/apps/saved_reports/tasks.py [hunk 7]
```diff
@@ -145,20 +147,29 @@ def send_email_report(self, recipient_emails, domain, report_slug, report_type,
...
         for recipient in recipient_emails:
             send_HTML_email(subject, recipient,
                             body, email_from=settings.DEFAULT_FROM_EMAIL,
                             smtp_exception_skip_list=LARGE_FILE_SIZE_ERROR_CODES)
 
     except Exception as er:
-        if getattr(er, 'smtp_code', None) in LARGE_FILE_SIZE_ERROR_CODES:
+        notify_exception(
+            None,
+            message="Encountered error while sending email",
```
Comment (esoergel) [inline corehq/apps/saved_reports/tasks.py]: or while generating the report

File: corehq/ex-submodules/couchexport/export.py [hunk 8]
```diff
@@ -313,19 +316,28 @@ def get_data(self):
...
+                wrapped_rows = name, (cls(row) for row in _wrap_row(first_entry, rows_to_wrap))
+            else:
+                rows = _wrap_row(rows[0], list(rows))
+                wrapped_rows.append(
+                    (name, [cls(row) for row in rows])
```
Comment (esoergel) [inline corehq/ex-submodules/couchexport/export.py]: Doesn't this evaluate all rows in memory?

**Comment by pr33thi** [comment]
@esoergel thanks for the review! fyi, I'm going to be pushing changes to try to get a component of this to work on my limited release of icds, and then I'll go back and address your comments (to explain all the emails you'll probably get)

File: corehq/apps/reports/standard/monitoring.py [hunk 9]
```diff
@@ -1792,22 +1801,83 @@ def _total_row(self, rows, report_data):
...
+            self.total_row = self.sum_rows_together(self.total_row, this_row)
+            print('(PV) monitoring 5')
+            for row in rows:
+                yield row
+            first_id_in_chunk = first_id_in_chunk + chunk_size
+        print('(PV) monitoring 6')
+        self.total_row = self.format_total_row(self.total_row)
+        print('(PV) monitoring 7')
+        yield self.total_row
+        print('==========')
```
[the remark this answers is not in the document]
Comment (kaapstorm) [reply corehq/apps/reports/standard/monitoring.py]: :nerd_face: You can shut Stickler up with
```python
from __future__ import print_function
```

[the remark this answers is not in the document]
Comment (pr33thi) [reply corehq/apps/reports/standard/monitoring.py]: haha good tip, I do want to make sure I got all the print statements out though, they were just for debugging. Sorry for all the emails!

File: corehq/apps/reports/standard/monitoring.py [hunk 10]
```diff
@@ -1792,22 +1801,74 @@ def _total_row(self, rows, report_data):
...
+    def get_all_rows(self):
+        user_es_query = EMWF.user_es_query(
+            self.domain, self.request.GET.getlist(EMWF.slug), self.request.couch_user
+        )
+        chunk_size = 10000
+        first_id_in_chunk = 0
+        user_chunk_total = user_es_query.fields(util.SimplifiedUserInfo.ES_FIELDS).size(chunk_size).run().total
+        while first_id_in_chunk < user_chunk_total:
+            user_chunk = user_es_query.fields(util.SimplifiedUserInfo.ES_FIELDS).start(first_id_in_chunk).size(chunk_size).run().hits
+            users = [util._report_user_dict(user) for user in user_chunk]
```
[the remark this answers is not in the document]
Comment (pr33thi) [reply corehq/apps/reports/standard/monitoring.py]: Thanks! https://github.com/dimagi/commcare-hq/pull/23782/commits/d398726f49682986d3f2adfafa47f67f67849f76

File: corehq/apps/reports/standard/monitoring.py (hunk 3 above)
Comment (pr33thi) [reply corehq/apps/reports/standard/monitoring.py]: Included in this commit: https://github.com/dimagi/commcare-hq/pull/23782/commits/d398726f49682986d3f2adfafa47f67f67849f76

File: corehq/apps/reports/standard/monitoring.py (hunk 5 above)
Comment (pr33thi) [reply corehq/apps/reports/standard/monitoring.py]: Included in this commit: https://github.com/dimagi/commcare-hq/pull/23782/commits/d398726f49682986d3f2adfafa47f67f67849f76

File: corehq/apps/saved_reports/tasks.py (hunk 6 above)
Comment (pr33thi) [reply corehq/apps/saved_reports/tasks.py]: https://github.com/dimagi/commcare-hq/pull/23782/commits/813d9daea54cdd41706bc657cde6dcfb31db2d02

File: corehq/apps/saved_reports/tasks.py (hunk 7 above)
Comment (pr33thi) [reply corehq/apps/saved_reports/tasks.py]: https://github.com/dimagi/commcare-hq/pull/23782/commits/d922ede67c934774c2b5c7c284b972c708102ae4

File: corehq/ex-submodules/couchexport/export.py (hunk 8 above)
Comment (pr33thi) [reply corehq/ex-submodules/couchexport/export.py]: Yes- that's ok with the chunking logic

File: corehq/apps/reports/standard/monitoring.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (pr33thi) [reply corehq/apps/reports/standard/monitoring.py]: https://github.com/dimagi/commcare-hq/pull/23782/commits/d398726f49682986d3f2adfafa47f67f67849f76

File: corehq/apps/reports/standard/monitoring.py [hunk 11]
```diff
@@ -1792,22 +1801,74 @@ def _total_row(self, rows, report_data):
...
                     all_users.add(user['user_id'])
-            total_row[3] = '%s / %s' % (len(active_users), len(all_users))
+            total_row[3] = {'numerator': len(active_users), 'demoninator': len(all_users)}
         else:
-            total_row[3] = '%s / %s' % (num, len(rows))
+            total_row[3] = {'numerator': num, 'denominator': len(rows)}
         return total_row
 
+    @property
+    def get_all_rows(self):
```
[the remark this answers is not in the document]
Comment (pr33thi) [reply corehq/apps/reports/standard/monitoring.py]: Good point, I changed it here so that it only uses the old logic when `self.view_by_groups` is true: https://github.com/dimagi/commcare-hq/pull/23782/commits/a2290d9a82232dacc981120eda4f1e2a1b0eaf9d

File: corehq/apps/reports/standard/monitoring.py (hunk 2 above)
Comment (pr33thi) [reply corehq/apps/reports/standard/monitoring.py]: This is addressed in this commit: https://github.com/dimagi/commcare-hq/pull/23782/commits/a2290d9a82232dacc981120eda4f1e2a1b0eaf9d

File: corehq/ex-submodules/couchexport/export.py (hunk 8 above)
Comment (esoergel) [reply corehq/ex-submodules/couchexport/export.py]: I do think it could be written to disk without evaluating in memory, by keeping it a generator even when writing the rows to the file.  Actually, not sure whether that's possible with our excel library, though I think it is with the csv lib.  Anyways, something to explore if this falls over with the full data set.

Comment (pr33thi) [reply corehq/ex-submodules/couchexport/export.py]: Totally- that is something I briefly looked into but had trouble doing with our excel library, so I tested this implementation at icds scale on my limited release and it worked fine, so I'll opt to save that as future work if needed

**Comment by dannyroberts** [comment]
Not sure if this is on your radar @pr33thi, but the build has been failing with

```
======================================================================
ERROR: couchexport.tests.test_writers:Excel2003ExportWriterTests.test_data_length
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/ex-submodules/couchexport/tests/test_writers.py", line 145, in test_data_length
    export_from_tables(tables, file_, format_)
  File "/mnt/commcare-hq-ro/corehq/ex-submodules/couchexport/export.py", line 35, in export_from_tables
    worksheet_title, row_generator = FormattedRow.wrap_all_rows(tables)
ValueError: need more than 1 value to unpack
----------------------------------------------------------------------
```
since 67cd139. Might want to see if you can debug that locally with `./manage.py test couchexport.tests.test_writers:Excel2003ExportWriterTests.test_data_length`.

**Comment by pr33thi** [comment]
@esoergel that makes sense- here's my solution: https://github.com/dimagi/commcare-hq/pull/23782/commits/e0aee6cab4038d9741d0d3f016f236408f82cf59

**Comment by pr33thi** [comment]
Thanks @dannyroberts!

**Review by esoergel — APPROVED** [review]
Looks good, thanks!

**Comment by pr33thi** [comment]
thanks @esoergel! I made one more commit that could use also review (just removed a bunch of the generator logic now that I'm chunking by a bunch of users at a time instead)
I'm also testing that in my release to make sure it doesn't break anything

**Review by calellowitz — APPROVED** [review]

File: corehq/ex-submodules/couchexport/export.py [hunk 12]
```diff
@@ -338,6 +338,7 @@ def _wrap_row(first_entry, rows):
                 first_entry = next(first_entry_rows)
                 wrapped_rows = name, (cls(row) for row in _wrap_row(first_entry, rows_to_wrap))
             else:
+                rows = list(rows)
                 rows = _wrap_row(rows[0], list(rows))
```
Comment (esoergel) [inline corehq/ex-submodules/couchexport/export.py]: nit: you can drop the second call to `list(rows)`

File: corehq/ex-submodules/couchexport/export.py [hunk 13]
```diff
@@ -313,19 +329,30 @@ def get_data(self):
...
                 rows = [rows]
-            ret.append(
-                (name, [cls(row) for row in rows])
-            )
-        return ret
+            return rows
+
+        wrapped_rows = []
+        for name, rows in tables:
+            if isinstance(rows, itertools.chain):
```
Comment (esoergel) [inline corehq/ex-submodules/couchexport/export.py]: similarly here I think it would be simpler to implement if you coerced to a generator up front rather than having parallel implementations

File: corehq/apps/reports/standard/monitoring.py [hunk 14]
```diff
@@ -1820,8 +1820,11 @@ def user_rows(self):
             self.domain, self.request.GET.getlist(EMWF.slug), self.request.couch_user
         )
         chunk_size = 50000
-        user_iterator = user_es_query.size(chunk_size).scroll()
-        for user_chunk in chunked(user_iterator, chunk_size):
+        first_id_in_chunk = 0
+        user_chunk_total = user_es_query.fields(util.SimplifiedUserInfo.ES_FIELDS).size(chunk_size).run().total
+        while first_id_in_chunk < user_chunk_total:
+            user_chunk = user_es_query.fields(util.SimplifiedUserInfo.ES_FIELDS).start(first_id_in_chunk).size(
+                chunk_size).run().hits
```
Comment (esoergel) [inline corehq/apps/reports/standard/monitoring.py]: The problem with this is that if it takes a long time to process, users could be added or removed, which would throw off these indices, possibly causing this to skip users or include duplicates.  Scroll queries avoid this problem.  The same issue occurred for exports: https://github.com/dimagi/commcare-hq/pull/20282

You could try using `iter_es_docs_from_query`, which uses a scroll query to get a list of ids up front, streams that list of IDs to disk (in case it's too big to hold in memory), and then pulls the documents in chunks as needed.  Individual documents may change throughout the duration of the query, but the set of documents returned should be fixed.

File: corehq/ex-submodules/couchexport/export.py [hunk 15]
```diff
@@ -31,10 +32,25 @@ def get_writer(format):
...
+            # the generator
+            row_generator_1, row_generator_2 = itertools.tee(row_generator)
+            header_table.append((worksheet_title, [next(row_generator_1)]))
+            row_table.append((worksheet_title, row_generator_2))
+    else:
+        header_table = tables
+        row_table = tables
+
+    writer.open(header_table, file, max_column_size=max_column_size)
+    writer.write(row_table, skip_first=True)
```
[the remark this answers is not in the document]
Comment (pr33thi) [reply corehq/ex-submodules/couchexport/export.py]: Great explanation: https://github.com/dimagi/commcare-hq/pull/23782/commits/a7ced9e5d163a0643e1b9498b35cc3ab31192110

File: corehq/ex-submodules/couchexport/export.py [hunk 16]
```diff
@@ -31,10 +32,25 @@ def get_writer(format):
...
 def export_from_tables(tables, file, format, max_column_size=2000):
-    tables = FormattedRow.wrap_all_rows(tables)
     writer = get_writer(format)
-    writer.open(tables, file, max_column_size=max_column_size)
-    writer.write(tables, skip_first=True)
+    tables = FormattedRow.wrap_all_rows(tables)
+
+    if isinstance(tables[0][1], itertools.chain) or isinstance(tables[0][1], types.GeneratorType):
+        header_table = []
+        row_table = []
```
[the remark this answers is not in the document]
Comment (pr33thi) [reply corehq/ex-submodules/couchexport/export.py]: Included in https://github.com/dimagi/commcare-hq/pull/23782/commits/a7ced9e5d163a0643e1b9498b35cc3ab31192110

File: corehq/ex-submodules/couchexport/export.py (hunk 12 above)
Comment (pr33thi) [reply corehq/ex-submodules/couchexport/export.py]: https://github.com/dimagi/commcare-hq/pull/23782/commits/72192e89865204e769395647f278d0662b6f8b07

File: corehq/ex-submodules/couchexport/export.py (hunk 13 above)
Comment (pr33thi) [reply corehq/ex-submodules/couchexport/export.py]: https://github.com/dimagi/commcare-hq/pull/23782/commits/93eccfad3abfe1b80ce79d9bd34e7e27b1f197f7

File: corehq/apps/reports/standard/monitoring.py (hunk 14 above)
Comment (pr33thi) [reply corehq/apps/reports/standard/monitoring.py]: Implemented here: https://github.com/dimagi/commcare-hq/pull/23782/commits/a900e5f50299dd463958631ef0d14dfaf61772da

File: corehq/ex-submodules/couchexport/export.py [hunk 17]
```diff
@@ -313,19 +322,24 @@ def get_data(self):
...
+            return rows
+
+        wrapped_rows = []
+        for name, rows in tables:
+            rows = itertools.chain([], rows)
+            first_entry_rows, rows_to_wrap = itertools.tee(rows)
+            first_entry = next(first_entry_rows)
+            wrapped_rows.append(
+                (name, [cls(row) for row in _wrap_row(first_entry, rows_to_wrap)]))
+        return wrapped_rows
```
Comment (esoergel) [inline corehq/ex-submodules/couchexport/export.py]: I'm still not crazy about this method.  I find it a little hard to follow.  I like your idea of extracting `_wrap_row`, but I'd pull the `first_entry` business into it too, since that's the only place it's needed.  I'd also avoid the usage of `tee`, since it [stores the list in memory when needed](https://docs.python.org/2/library/itertools.html#itertools.tee), (although the compiler may be smart enough to see that `first_entry_rows` isn't accessed again, not sure).  Lastly, you can also use a generator expression rather than a list expression and avoid evaluating the rows here.

I realize some of those suggestions are kind of vague, so here's my attempt:

```python
def coerce_to_list_of_rows(rows):
    rows = iter(rows)
    # peek at the first element, then add it back
    first_entry = next(rows)
    rows = itertools.chain([first_entry], rows)
    if first_entry and (not hasattr(first_entry, '__iter__') or isinstance(first_entry, six.string_types)):
        soft_assert_type_text(first_entry)
        # `rows` is actually just a single row, so wrap it
        return [rows]
    return rows

return [
    (name, (cls(row) for row in coerce_to_list_of_rows(rows)))
    for name, rows in tables
]
```

**Review by esoergel — COMMENTED** [review]
Looking good, thanks for your patience with my reviews!  Couple last little comments

File: corehq/apps/reports/standard/monitoring.py [hunk 18]
```diff
@@ -716,12 +716,12 @@ def rows(self):
...
                           for i, col in enumerate(row[1:])]
-                rows.append(row)
+                yield row
             else:
-                rows.append([self.get_user_link(simplified_user), '--'])
+                yield [self.get_user_link(simplified_user), '--']
         if self.all_relevant_forms:
             self.total_row = [_("All Users")] + totals
-        return rows
+        yield self.total_row
```
[the remark this answers is not in the document]
Comment (pr33thi) [reply corehq/apps/reports/standard/monitoring.py]: https://github.com/dimagi/commcare-hq/pull/23782/commits/0e27cbad615364da5382b20504571ab07dcd186c

File: corehq/ex-submodules/couchexport/export.py (hunk 17 above)
Comment (pr33thi) [reply corehq/ex-submodules/couchexport/export.py]: https://github.com/dimagi/commcare-hq/pull/23782/commits/a2ff0cb8e39658671d727969d4ba6302c76ba343

**Review by esoergel — APPROVED** [review]
